Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp

```diff
@@ -76,6 +76,34 @@ class TIFFLoadingContext {
         BigEndian,
     };
 
+    static ErrorOr<u8> read_component(BigEndianInputBitStream& stream, u8 bits)
+    {
+        // FIXME: This function truncates everything to 8-bits
+        auto const value = TRY(stream.read_bits<u32>(bits));
+
+        if (bits > 8)
+            return value >> (bits - 8);
+        return value << (8 - bits);
```

Comment:
This is a low-quality way of doing this. Imagine the `bits` being 2. You have 4 values, 0, 1, 2, 3. This maps them to 0, 64, 128, 192. Instead, you want to repeat the bits so they fill  the 8 bits: 0 still goes to 0, 1 goes to 0b01010101, 2 goes to 0b10101010, 3 goes to 0b11111111 – or, in decimal, 0, 85, 170, 255. Note how these are equidistant, and map the highest value (3) to the highest 8-bit value (255) – white stays white.

If you only need to handle bit depths that divide 8 (1, 2, 4), the standard trick to repeat the bits is to do `value * (255 / ((1 << bits) - 1))`, but I think it doesn't do the right thing for e.g. 6 bpp. (Per tiff spec, only 4 bpp are supported, so that might be ok.)

It's important to only do this if the bits are actual colors – if they're palette indices, you have to keep them as they are of course.

Reply:
Oh, it never crossed my mind, thanks for your comment!

> Per tiff spec, only 4 bpp are supported, so that might be ok.

I don't think that it's actually specified in the general case. For grayscale, only 4 and 8 are supported, but here is a quote (page 30 in the description of `Compression`) that seems to indicate that 6bpp are supported (while already not recommended in the 90s).

> If the number of bits per component is not a power of 2, and you are willing to give up some space for better performance, use the next higher power of 2. For example, if your data can be represented in 6 bits, set BitsPerSample to 8 instead of 6, and then convert the range of the values from [0,63] to [0,255].


> It's important to only do this if the bits are actual colors – if they're palette indices, you have to keep them as they are of course.

We can see experience talking here :laughing: